clients/src/main/java/org/apache/kafka/common/telemetry/internals/ClientTelemetrySender.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.common.telemetry.internals;
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.requests.AbstractRequest.Builder;
+import org.apache.kafka.common.requests.GetTelemetrySubscriptionsResponse;
+import org.apache.kafka.common.requests.PushTelemetryResponse;
+
+import java.util.Optional;
+
+/**
+ * The interface used by the `NetworkClient` to send telemetry requests.
+ */
+public interface ClientTelemetrySender extends AutoCloseable {
+
+    /**
+     * Return the next time when the telemetry API should be attempted (i.e., interval time has elapsed).
+     * <p>
+     * If another telemetry API is in-flight, then {@code timeoutMs} should be returned as the
+     * maximum wait time.
+     *
+     * @param timeoutMs The timeout for the inflight telemetry API call.
+     * @return remaining time in ms till the telemetry API be attempted again.
+     */
+    long timeToNextUpdate(long timeoutMs);
+
+    /**
+     * Return the telemetry request based on client state i.e. determine if
+     * {@link org.apache.kafka.common.requests.GetTelemetrySubscriptionsRequest} or
+     * {@link org.apache.kafka.common.requests.PushTelemetryRequest} be constructed.
+     *
+     * @return request for telemetry API call.
+     */
+    Optional<Builder<?>> createRequest();
+
+    /**
+     * Handle successful response for get telemetry subscriptions request.
+     *
+     * @param response subscriptions telemetry API response
+     */
+    void handleResponse(GetTelemetrySubscriptionsResponse response);
+
+    /**
+     * Handle successful response for push telemetry request.
+     *
+     * @param response push telemetry API response
+     */
+    void handleResponse(PushTelemetryResponse response);
+
+    /**
+     * Handle response for failed get telemetry subscriptions request.
+     *
+     * @param kafkaException the fatal exception.
+     */
+    void handleFailedGetTelemetrySubscriptionsRequest(KafkaException kafkaException);
```

Comment:
`handleFailedGetTelemetricSubscriptionResponse` ? We are talking about the response right?

Can it take a Throwable type instead of KafkaException? The exception can be retriable as well right?

Reply:
The only way network client fails a Request is really NetworkException I think, which is handled in the networkClient.

Reply:
It's handling of failed request, but yes the exception arrived in response. I tried to be more consistent as method names defined in other packages for example: https://github.com/apache/kafka/blob/c81a7252195261f649faba166ee723552bed4d81/clients/src/main/java/org/apache/kafka/clients/MetadataUpdater.java#L78

Reply:
> Can it take a Throwable type instead of KafkaException?

Sorry I didn't understand the reasoning behind the suggestion, please help me? Isn't `Throwable` more generic.

> network client fails a Request is really NetworkException I think

I have seen UnsupportedVersionException coming through.

Reply:
Though the behaviour of failed request handling would be to retry automatically after elapsed interval unless we found an exception which suggests to stop re-trying.

I still do not get how `Throwable` will solve the problem of retriable exceptions. Are there some other retriable exceptions which do not extend `RetriableException`? I might be missing something here, can you please help.